provider/macros/src/lib.rs

```diff
@@ -92,12 +92,167 @@ mod tests;
 /// implement it on the markers.
 pub fn data_struct(attr: TokenStream, item: TokenStream) -> TokenStream {
     TokenStream::from(data_struct_impl(
-        parse_macro_input!(attr as AttributeArgs),
+        parse_macro_input!(attr as DataStructArgs),
         parse_macro_input!(item as DeriveInput),
     ))
 }
 
-fn data_struct_impl(attr: AttributeArgs, input: DeriveInput) -> TokenStream2 {
+pub(crate) struct DataStructArgs {
+    args: Punctuated<DataStructArg, Token![,]>,
+}
+
+impl Parse for DataStructArgs {
+    fn parse(input: ParseStream<'_>) -> parse::Result<Self> {
+        let args = input.parse_terminated(DataStructArg::parse, Token![,])?;
+        Ok(Self { args })
+    }
+}
+struct DataStructArg {
+    marker_name: Path,
+    key_lit: Option<LitStr>,
+    fallback_by: Option<LitStr>,
+    extension_key: Option<LitStr>,
+    fallback_supplement: Option<LitStr>,
+}
+
+impl DataStructArg {
+    fn new(marker_name: Path) -> Self {
+        Self {
+            marker_name,
+            key_lit: None,
+            fallback_by: None,
+            extension_key: None,
+            fallback_supplement: None,
+        }
+    }
+}
+
+impl Parse for DataStructArg {
+    fn parse(input: ParseStream<'_>) -> parse::Result<Self> {
+        let path: Path = input.parse()?;
+
+        fn at_most_one_option<T>(
+            o: &mut Option<T>,
+            new: T,
+            name: &str,
+            span: Span,
+        ) -> parse::Result<()> {
+            if o.replace(new).is_some() {
+                Err(parse::Error::new(
+                    span,
+                    format!("marker() cannot contain multiple {name}s"),
+                ))
+            } else {
+                Ok(())
+            }
+        }
+
+        if path.is_ident("marker") {
+            let content;
+            let _paren = parenthesized!(content in input);
+            let mut marker_name: Option<Path> = None;
+            let mut key_lit: Option<LitStr> = None;
+            let mut fallback_by: Option<LitStr> = None;
+            let mut extension_key: Option<LitStr> = None;
+            let mut fallback_supplement: Option<LitStr> = None;
+            let punct = content.parse_terminated(DataStructMarkerArg::parse, Token![,])?;
```

Comment:
Observation: it seems `content.parse_terminated` is one way to force the entire stream to be consumed, so that there are no extra tokens.

Reply:
Yep. In general most _toplevel_ parsing functions (e.g. `parse_args()`, `parse()`, etc) ensure this too. You rarely need to think about this when writing `syn` parser code.